grails-app/services/org/pih/warehouse/core/LocationService.groovy

```diff
@@ -718,4 +724,83 @@ class LocationService {
         }
         existingLocation.delete(flush: true)
     }
+
+    Location createOrUpdateLocation(Map params) {
+        Location location
+        if (params.id) {
+            location = Location.findById(params.id)
+        } else if (params.locationNumber) {
+            location = Location.findByNameOrLocationNumber(params.name, params.locationNumber)
+        } else {
+            location = Location.findByName(params.name)
+        }
+        if (!location) {
+            location = new Location()
+        }
+
+        location.name = params.name
+        location.active = Boolean.valueOf(params.active)
+        location.locationNumber = params.locationNumber
+        location.locationGroup = params.locationGroup ? LocationGroup.findByName(params.locationGroup) : null
+        location.parentLocation = params.parentLocation ? Location.findByNameOrLocationNumber(params.parentLocation, params.parentLocation) : null
+        location.organization = params.organization ? Organization.findByCodeOrName(params.organization, params.organization) : null
```

Comment:
couldn't we do such binding through command/dto object? So to create some kind of dto that would accept:
```groovy
class LocationCreateDto {
  String locationNumber
  LocationGroup locationGroup
  ...
}
```
and to think about a way to map the dto to entity (Location)?

Maybe this is not good time to do this (I assume you've copied the existing code), but wanted to raise that.
I just think this method has a bunch of unnecessary "boilerplate code"

Reply:
@jmiranda to be clear, I meant even something like:
```groovy
Location createOrUpdateLocation(LocationCreateDto locationDto) {
  ...
}
```
but this would have to happen at the controller level which as I wrote in on of the comments when I realized it - it might be difficult to achieve with the current "generic" implementation of the DataService methods. 

I agree that the `findByName` cases done for `locationGroup` would be difficult to achieve (unless if we maybe overwrote the setter for the particular dto), but my point was to just maybe try to bind the "common" part, and the rest, that requires some logic, like `locationGroup`, `parentLocation` to be still done manually.
It was just a "loose thought" that I wish we followed this dto/command approach in the future in the places where it is possible.